### PR TITLE
Cleanup and Documentation

### PR DESCRIPTION
**Pull request by rhafer**
This adds documentation about the new public cloud related additions and cleans up a few minor things.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,6 +1,7 @@
-# container-suseconnect [![Build Status](https://travis-ci.org/SUSE/container-suseconnect.svg?branch=master)](https://travis-ci.org/SUSE/container-suseconnect) [![GoDoc](https://godoc.org/github.com/SUSE/container-suseconnect?status.png)](https://godoc.org/github.com/SUSE/container-suseconnect)
+# container-suseconnect [![Build Status](https://travis-ci.org/SUSE/container-suseconnect.svg?branch=master)](https://travis-ci.org/SUSE/container-suseconnect) [![BoDoc](https://godoc.org/github.com/SUSE/container-suseconnect?status.png)](https://godoc.org/github.com/SUSE/container-suseconnect)
```
Comment (flavio) [inline README.md:1]: Why did you rename from `GoDoc` to `BoDoc`? Is that a mistake?

File: README.md [hunk 2]
```diff
@@ -143,6 +144,51 @@ outlined in the [Building images on non SLE distributions](#building-images-on-n
 section.
 
 
+### Building images on-demand SLE instances in the public cloud
+
+When building container images on SLE instance that where launched as so-called
```
Comment (flavio) [inline README.md:149]: `on SLE instances that were launched`

File: README.md [hunk 3]
```diff
@@ -143,6 +144,51 @@ outlined in the [Building images on non SLE distributions](#building-images-on-n
...
+"on-demand" or "pay as you go" instances on a Public Cloud (as AWS, GCE or Azure)
+some additional steps have to be performed.
+
+For installing packages and updates the "on-demand" public cloud instance are
+connected to a public cloud specific update infrastructure which is based around
+some RMT servers operated by SUSE on the various Public Cloud Providers.
```
Comment (flavio) [inline README.md:155]: ~some~

File: README.md [hunk 4]
```diff
@@ -143,6 +144,51 @@ outlined in the [Building images on non SLE distributions](#building-images-on-n
...
+
+To be able access this update infrastructure the instances need to perform some
```
Comment (flavio) [inline README.md:157]: ~some~

File: README.md [hunk 5]
```diff
@@ -143,6 +144,51 @@ outlined in the [Building images on non SLE distributions](#building-images-on-n
...
+additional steps to locate the required service and authenticate with them.
+(Some more details on this are outline in a
+[Blog Series on suse.com starting here](https://suse.com/c/a-new-update-infrastructure-for-the-public-cloud/).
```
Comment (flavio) [inline README.md:160]: I would drop the `(` and `)`. But that's a matter of personal taste.

File: README.md [hunk 6]
```diff
@@ -143,6 +144,51 @@ outlined in the [Building images on non SLE distributions](#building-images-on-n
...
+
+In order to be able to build containers on this type of instances a new service
```
Comment (flavio) [inline README.md:162]: `In order to build containers`

I make the same "mistake" too :)

File: README.md [hunk 7]
```diff
@@ -143,6 +144,51 @@ outlined in the [Building images on non SLE distributions](#building-images-on-n
...
+was introduced, that service is called "containerbuild-regionsrv" and will be
```
Comment (flavio) [inline README.md:163]: Replace `"` with `

File: README.md [hunk 8]
```diff
@@ -143,6 +144,51 @@ outlined in the [Building images on non SLE distributions](#building-images-on-n
...
+systemctl start containerbuild-regionsrv
+```
+
+In order to have it started automatically (e.g. after reboot) please use:
+
+```
+systemctl enable  containerbuild-regionsrv
+```
+
+The zypper plugins provided by container-suseconnect will then connect to this
```
Comment (flavio) [inline README.md:178]: Wrap container-suseconnect with `

File: README.md [hunk 9]
```diff
@@ -143,6 +144,51 @@ outlined in the [Building images on non SLE distributions](#building-images-on-n
...
+service for getting authentication details and information about which update
+server to talk to. In order for that to work the container as to be build with
```
Comment (flavio) [inline README.md:180]: > the container **h**as to be buil**t**

File: README.md [hunk 10]
```diff
@@ -143,6 +144,51 @@ outlined in the [Building images on non SLE distributions](#building-images-on-n
...
+host networking enabled. I.e. you need to call `docker build` with `--network host`:
+
+```
+docker build --network host <builddir>
+```
+
+As the update infrastructure in the Public Clouds is based upon RMT the same
```
Comment (flavio) [inline README.md:187]: This is just my personal taste, feel free to ignore:

> The update infrastructure in the Public Clouds is based upon RMT hence the same

File: vendor/modules.txt [hunk 11]
```diff
@@ -1,5 +1,7 @@
 # github.com/cpuguy83/go-md2man/v2 v2.0.0-20190314233015-f79a8a8ca69d
 github.com/cpuguy83/go-md2man/v2/md2man
+# github.com/mssola/capture v0.0.0-20180102083101-24891570e9b2
+github.com/mssola/capture
```
Comment (flavio) [inline vendor/modules.txt:4]: I know this wasn't probably generated by you. I wonder why the other repo is left there commented. Have you tried to run a `go mod tidy` to see if the file is cleaned up?

File: README.md (hunk 1 above)
Comment (rhafer) [reply README.md:1]: Oh wow. I guess that was just me being fat-fingered. Thanks for finding.

File: vendor/modules.txt (hunk 11 above)
Comment (rhafer) [reply vendor/modules.txt:4]: That file is generated by `go mod vendor` and I think it's its default behavior to add a comment the with the repo, version, timestamp and commit id of the repo and then the list of go modules from that repository.
Or are you referring to something different?

File: README.md (hunk 4 above)
Comment (mssola) [reply README.md:157]: `To be able to access`. @flavio I'd actually leave the `some` in this case.

File: README.md [hunk 12]
```diff
@@ -143,6 +144,51 @@ outlined in the [Building images on non SLE distributions](#building-images-on-n
...
+"on-demand" or "pay as you go" instances on a Public Cloud (as AWS, GCE or Azure)
+some additional steps have to be performed.
+
+For installing packages and updates the "on-demand" public cloud instance are
+connected to a public cloud specific update infrastructure which is based around
+some RMT servers operated by SUSE on the various Public Cloud Providers.
+
+To be able access this update infrastructure the instances need to perform some
+additional steps to locate the required service and authenticate with them.
+(Some more details on this are outline in a
```
Comment (mssola) [inline README.md:159]: `are outlined`

File: README.md (hunk 10 above)
Comment (mssola) [reply README.md:187]: Personally I'd go with something like: `Since the update infrastructure in the Public Clouds is based on RMT, the same` (notice the comma after RMT).

File: README.md [hunk 13]
```diff
@@ -143,6 +144,51 @@ outlined in the [Building images on non SLE distributions](#building-images-on-n
...
+
+```
+docker build --network host <builddir>
+```
+
+As the update infrastructure in the Public Clouds is based upon RMT the same
+restrictions with regard to build SLE images for SLE versions differing from the
+SLE version of the host apply here as well. (See above)
+
+
```
Comment (mssola) [inline README.md:191]: Remove the extra blank line (there are two now)

File: internal/regionsrv/server.go [hunk 14]
```diff
@@ -36,28 +36,28 @@ type SuseBuildConfig struct {
...
-// server. You may tweak this by providing `SUSE_BUILD_IP` and/or
-// `SUSE_BUILD_PORT`, otherwise `0.0.0.0` and `7956` are taken as defaults for
+// containerBuildSrvAddress returns a string containing the full address of the TCP
+// server. You may tweak this by providing `CONTAINER_BUILD_IP` and/or
+// `CONTAINER_BUILD_PORT`, otherwise `0.0.0.0` and `7956` are taken as defaults for
 // the IP and the port respectively.
-func suseBuildAddress() string {
-	ip := os.Getenv("SUSE_BUILD_IP")
+func containerBuildSrvAddress() string {
+	ip := os.Getenv("CONTAINER_BUILD_IP")
```
Comment (mssola) [inline internal/regionsrv/server.go:44]: I can understand the general renaming of it all, by I'm not 100% sold on the idea of changing the environment variables from `SUSE_` to `CONTAINER_`. I like the former better because it specifies that it's something SUSE-specific. Whereas leaving it as `CONTAINER_` might confuse customers thinking that this environment variable is coming from Docker or somewhere else.

Comment (rhafer) [reply internal/regionsrv/server.go:44]: How strong is your opinion on this? Can we agree to disagree here?  :)

I'd really like to have it consistent with the naming in the service. Also I am not a huge fan of spreading the company name into variable names, might be from the experience of previous acquisitions. I admit this point is a bit mood for a tool call `container-suseconnect` ;-)

File: README.md (hunk 5 above)
Comment (rhafer) [reply README.md]: done

File: README.md (hunk 6 above)
Comment (rhafer) [reply README.md]: :heavy_check_mark:

File: README.md (hunk 10 above)
Comment (rhafer) [reply README.md]: Yeah, that sounds even better. Thanks.

File: internal/regionsrv/server.go (hunk 14 above)
Comment (mssola) [reply internal/regionsrv/server.go:44]: I have no strong opinions on this. Go as you prefer, it was just a suggestion :)

**Comment by rhafer** [comment]
Hm, not sure what it is yet. But it seem the `TestReadConfigFromServerBadResponse` is flaky sometimes. See: https://travis-ci.org/github/SUSE/container-suseconnect/jobs/700008966

Or maybe it's just travis that had a hickup.

**Comment by mssola** [comment]
> Hm, not sure what it is yet. But it seems the TestReadConfigFromServerBadResponse is flaky. See: https://travis-ci.org/github/SUSE/container-suseconnect/jobs/700008966

That's weird. Anyways, it should be pretty easy to workaround. It looks like the error has a different message (even though they both refer to a parsing error). That might be about the Go version or something like that. I'd say that just checking that `err` is not nil https://github.com/SUSE/container-suseconnect/blob/master/internal/regionsrv/server_test.go#L112 should be enough.

**Comment by rhafer** [comment]
> That might be about the Go version or something like that.

Hm, doesn't seem so. Then it would fail consistently on that go version. There were failed runs on go1.12 (and -tip). Also, looking at the encoding/json sources I think the errror we're expecting is the right one. 

> I'd say that just checking that `err` is not nil https://github.com/SUSE/container-suseconnect/blob/master/internal/regionsrv/server_test.go#L112 should be enough.

I'd rather not do that for now. We might be papering over some real errors with that. I now re-ran the tests quite a few times and they always succeeded. I think I'd blame the failures on a glitch in the travis workers for now.

Let's merge this PR as is and keep an I on failing tests in the future.

File: vendor/modules.txt (hunk 11 above)
Comment (flavio) [reply vendor/modules.txt:4]: No worries, it that's how go mod vendor works. Fine with me :)

File: internal/regionsrv/server.go (hunk 14 above)
Comment (flavio) [reply internal/regionsrv/server.go:44]: Let's stick with a value that goes after the projects' name. Hence let's stick with the current code

**Comment by rhafer** [comment]
> I'd rather not do that for now. We might be papering over some real errors with that. I now re-ran the tests quite a few times and they always succeeded. I think I'd blame the failures on a glitch in the travis workers for now.

Turns out that there actually is a real problem in the code see #45 for details.
